src/sdk/pynni/nni/nas/pytorch/trainer.py

```diff
@@ -0,0 +1,83 @@
+import json
+import logging
+from abc import abstractmethod
+
+import torch
+
+from .base_trainer import BaseTrainer
+
+_logger = logging.getLogger(__name__)
+_logger.setLevel(logging.INFO)
+
+
+class TorchTensorEncoder(json.JSONEncoder):
+    def default(self, o):  # pylint: disable=method-hidden
+        if isinstance(o, torch.Tensor):
+            olist = o.tolist()
+            if "bool" not in o.type().lower() and all(map(lambda d: d == 0 or d == 1, olist)):
+                _logger.warning("Every element in %s is either 0 or 1. "
+                                "You might consider convert it into bool.", olist)
+            return olist
+        return super().default(o)
+
+
+class Trainer(BaseTrainer):
+    def __init__(self, model, mutator, loss, metrics, optimizer, num_epochs,
+                 dataset_train, dataset_valid, batch_size, workers, device, log_frequency, callbacks):
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu") if device is None else device
+        self.model = model
+        self.mutator = mutator
+        self.loss = loss
+
+        self.metrics = metrics
+        self.optimizer = optimizer
+
+        self.model.to(self.device)
+        self.mutator.to(self.device)
+        self.loss.to(self.device)
```

Comment:
If there are multiple GPUs on system, it seems only one GPU can be used due to these `to` statements

Reply:
Yes. Data/Model paralleled one-shot NAS is in future plan.